GS: Detect triangles-are-quads in primitive overlap

TJnotJT opened the pull request:
### Description of Changes
1. Added utility functions for detecting when vertices form axis-aligned right triangles or quads (original code by TellowKrinkle, modified by me).
2. Updated primitive overlap detection to account for pairs of triangles that together form axis-aligned quads.

### Rationale behind Changes
Currently, pairs of axis-aligned right triangles that form a quad are treated as overlapping. This PR adds logical to detect such triangles as non-overlapping. This should reduce barriers (OpenGL/Vulkan) and copies (DX11) when doing SW blending with multidraw (requires higher than basic blending).

In a dump with DX11/max blending, there were significant reductions in the number of copies. Many of these are menus but some are also in-game.

Examples:
```
Gradius V_SLUS-20712_20220830220750.gs.xz
Dark Cloud_SCUS-97111_20231026205150.gs.xz
kaan.gs.xz
Valkyrie_Profile_2_-_Silmeria_SLUS-21452_20230105135006.gs.xz
Jak_X_-_Combat_Racing_SCUS-97429_20240317114104.gs.xz
Legacy of Kain - Blood Omen 2_SLUS-20024_20221019151207.gs.xz
Magical Sports - Go Go Golf_SLPS-20037_20230731171544.gs.xz
Call_of_Duty_-_Finest_Hour_SLUS-20725_20230514111828.gs.xz
Grand_Theft_Auto_-_San_Andreas_SLUS-20946_20230725171446.gs.xz
Bloody Roar 3_SLUS-20212_20230307054849.gs.xz
```

In some dumps the number of copes in basic blend changed slightly (up or down). I have checked several by hand and they appear to be a result of increased accuracy with single-quad detection in this PR (single quads are special cases that are handled differently in both PR and master).

I also profiled the modified function `TrianglesAreQuads()` in both the PR and master and CPU overhead appears to be negligible (though I only checked one dump). Hopefully, there be no difference in games that do not require quad detection.

### Suggested Testing Steps
- Benchmarking games to make sure the performance did not drop and accuracy was not affected.
- So far I have done only dumps runs in DX11. However, all renderers should be run eventually.

### Did you use AI to help find, test, or implement this issue or feature?
I used AI to brainstorm deas and review my code. Nothing was copy/pasted directly in the final PR.

lightningterror commented on pcsx2/GS/GSUtil.h:
```diff
@@ -73,4 +74,283 @@ class GSUtil
+
+	// Determines ordering of two triangles in parallel if both are right.
+	// More efficient than calling IsTriangleRight twice.
+	template <u32 tme, u32 fst>
```
Would be better to move some of the stuff in the cpp file instead, same for the other functions below.

TJnotJT replied:
Good point -- it does seem a bit out of place in the header. I put it there to make it inlinable; otherwise, we’d lose some optimization, especially since it’s a templated function. That said, the original function was by @TellowKrinkle, so maybe they could chime in on whether it’s better, from an optimization standpoint, to keep it in the header or move it to the .cpp.

JordanTheToaster reviewed (approved):
Appears to work correctly large gains in specific games.

<img width="677" height="254" alt="Image" src="https://github.com/user-attachments/assets/867b86b8-a2d4-4b30-b695-94d767e3f9a3" />

<img width="671" height="269" alt="Image" src="https://github.com/user-attachments/assets/5729e3db-ff30-4d6a-bbab-1283f9b23824" />

<img width="677" height="257" alt="Image" src="https://github.com/user-attachments/assets/8070c2b9-bce1-4d0f-a4ca-bf56b531c8de" />

TJnotJT commented:
Good catch — I had taken a shortcut in the quad detection logic by assuming that if the first two pairs of triangles formed quads, then the remaining triangles in the draw were most likely quads as well. Unfortunately, that assumption doesn’t hold true in the dump above.

For context, the game is implementing a fog effect using triangles-as-quads, but some of the quads that overlap the scissor boundary are clipped into smaller triangles. This breaks the detection logic that assumes pairs of triangles always form a quad:
<img width="640" height="448" alt="frame_out" src="https://github.com/user-attachments/assets/7e6d36a4-1240-406b-af30-698927ea28b6" />

If we can detect this situation, we might still be speed up the scene considerably.

Converting to draft until this is figured out.

TJnotJT commented:
Changes:
1. Moved the triangle functions to GSState.cpp as per LT's recommendation.
2. Reworked the detection a bit so that it only happens a overlap computing time (should affect lower blend levels less CPU-wise).
3. Removed a couple shortcuts to address accuracy issues noted above by bigol83 and others I found in a dump run.

Still need to redo dump runs.

@bigol83 When you have the time could you please recheck GTA? The glitches should be fixed though the FPS may still be low on maximum blend. It seems like particle effects genuinely has a lot of overlap so there's only so much that we can do. Fortunately, the effect doesn't actually require maximum blending -- even full blending improves the FPS drastically.

bigol83 commented:
Issues with maximum blending has been solved, performance is the same, still around 11 fps

TJnotJT commented:
That's good, thank you for checking.

lightningterror reviewed (commented):
There's some games that also gain barriers as well, dump run was done using max blend.
Ace combat games being one of them, full list below:
[vk max blend.txt](https://github.com/user-attachments/files/22744890/vk.max.blend.txt)

TJnotJT commented:
Nice -- good to see it’s helped with some barriers as well.
